src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,134 @@
+                case ImmutableQueueGenericInterfaceTypeName:
+                    return  ImmutableQueueTypeName;
+                case ImmutableSortedSetGenericTypeName:
+                    return  ImmutableSortedSetTypeName;
+                case ImmutableHashSetGenericTypeName:
+                case ImmutableSetGenericInterfaceTypeName:
+                    return  ImmutableHashSetTypeName;
+                default:
+                    // TODO: Refactor exception throw following serialization exception changes. 
+                    throw new NotSupportedException(SR.Format(SR.DeserializeTypeNotSupported, immutableCollectionTypeName));
+            }
+        }
+
+        private string GetDelegateKey(
+            Type immutableCollectionType,
+            Type elementType,
+            out Type underlyingType,
+            out string constructingTypeName)
+        {
+            // Use the generic type definition of the immutable collection to determine an appropriate constructing type,
+            // i.e. a type that we can invoke the `CreateRange<elementType>` method on, which returns an assignable immutable collection.
+            underlyingType = immutableCollectionType.GetGenericTypeDefinition();
+            constructingTypeName = GetConstructingTypeName(underlyingType.FullName);
+
+            return $"{constructingTypeName}:{elementType.FullName}";
+        }
+
+        public void RegisterImmutableCollectionType(Type immutableCollectionType, Type elementType, JsonSerializerOptions options)
+        {
+            // Get a unique identifier for a delegate which will point to the appropiate CreateRange method.
+            string delegateKey = GetDelegateKey(immutableCollectionType, elementType, out Type underlyingType, out string constructingTypeName);
+
+            // Exit if we have registered this immutable collection type.
+            if (CreateRangeDelegates.ContainsKey(delegateKey))
+            {
+                return;
+            }
+
+            // Get the constructing type.
+            Type constructingType = underlyingType.Assembly.GetType(constructingTypeName);
+
+            // Create a delegate which will point to the CreateRange method.
+            object createRangeDelegate = options.ClassMaterializerStrategy.ImmutableCreateRange(constructingType, elementType);
+            Debug.Assert(createRangeDelegate != null);
+
+            // Cache the delegate
+            CreateRangeDelegates.TryAdd(delegateKey, createRangeDelegate);
+        }
+
+        public override IEnumerable CreateFromList(ref ReadStack state, IList sourceList, JsonSerializerOptions options)
+        {
+            Type immutableCollectionType = state.Current.JsonPropertyInfo.RuntimePropertyType;
+            Type elementType = state.Current.GetElementType();
+
+            string delegateKey = GetDelegateKey(immutableCollectionType, elementType, out _, out _);
+            Debug.Assert(CreateRangeDelegates.ContainsKey(delegateKey));
```

Comment:
Why is this guaranteed to be true here, when it could be false in `RegisterImmutableCollectionType`? Maybe add a comment?